webapp/src/app.tsx

```diff
@@ -1304,27 +1305,38 @@ export class ProjectView
         if (checkAsync)
             return checkAsync.then(() => this.openHome());
 
-        let doSync = false
+        let p = Promise.resolve();
         // check our multi-tab session
         if (workspace.isHeadersSessionOutdated()) {
              // reload header before loading
             pxt.log(`multi-tab sync before load`)
-            doSync = true;
-        }
-        // check our cloud sync
-        const RESYNC_TIME_SEC = 5 * 60 // 5 minutes
-        const headerCloudSyncOutdated = auth.loggedInSync()
-            && Util.nowSeconds() - workspace.getHeaderLastCloudSync(h) > RESYNC_TIME_SEC
-        if (headerCloudSyncOutdated) {
-            pxt.log(`cloud sync before load`)
-            doSync = true;
-        }
-
-        let p = Promise.resolve();
-        if (doSync) {
             p = p.then(() => workspace.syncAsync().then(() => { }))
         }
 
+        // Try a quick cloud fetch. If it doesn't complete within X second(s),
+        // continue on.
+        const timeout = Util.timeout(1500/*1.5 seconds*/)
+        p = p.then(() => {
+            return Promise.any([
+                timeout,
+                // start a partial cloud sync
+                cloud.syncAsync([h])
+                    .then((changes) => {
+                        if (changes.length) {
+                            if (timeout.isResolved()) {
+                                // We are too late; the editor has already been loaded.
+                                // Call the onChanges handler to update the editor.
+                                cloud.onChangesSynced(changes)
```

Comment:
Thinking about when we have a conflict resolution dialog available... can a conflict be detected here?

Reply:
Let's assume we did a minor update to the editor (so we automatically modify the pxt.json file with the new target version information) and the connection is slow (1 minute to sync), as soon as the editor will load the project will be "modified" and the local copy will win.
Do you merge changes with the cloud or do you blanket replace the version?

Reply:
> I would recommend fault injection (inject large delay semi-randomly) in any code path that talks to a cloud API. This will surface a lot of races naturally.

Yes this is important.

Reply:
Blanket replace. Yes, if you edit locally before the cloud has synced, you'll overwrite whatever is in the cloud. This is bad, and a dialog would help, but really we need history. We need git.